GoogleDataTransport/GDTCORLibrary/GDTCORDirectorySizeCalculator.m

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "GoogleDataTransport/GDTCORLibrary/Internal/GDTCORDirectorySizeCalculator.h"
+
+@interface GDTCORDirectorySizeCalculator ()
+
+/** The observed directory path. */
+@property(nonatomic, readonly) NSString *directoryPath;
+
+/** The cached content size of the observed directory. */
+@property(nonatomic, nullable) NSNumber *cachedSizeBytes;
+
+@end
+
+@implementation GDTCORDirectorySizeCalculator
+
+- (instancetype)initWithDirectoryPath:(NSString *)path {
+  self = [super init];
+  if (self) {
+    _directoryPath = path;
+  }
+  return self;
+}
+
+- (GDTCORStorageSizeBytes)directoryContentSize {
+  if (self.cachedSizeBytes == nil) {
+    self.cachedSizeBytes = @([self calculateDirectoryContentSize]);
+  }
+
+  return self.cachedSizeBytes.unsignedLongLongValue;
+}
+
+- (void)fileWithSize:(GDTCORStorageSizeBytes)fileSize wasAddedAtPath:(NSString *)path {
+  if (![path hasPrefix:self.directoryPath]) {
+    // Ignore because the file is not inside the directory.
+    return;
+  }
+
+  self.cachedSizeBytes = @([self directoryContentSize] + fileSize);
+}
+
+- (void)fileWithSize:(GDTCORStorageSizeBytes)fileSize wasRemovedAtPath:(NSString *)path {
```

Comment:
Both options (existing and proposed) read a bit awkward to me, but the current version reads a bit more like a sentence to me. @paulb777 @ryanwilson do you have an opinion on it?

Reply:
I like @visumickey's suggestion. It clears up the confusion I had when I first saw those two methods.  However, I might be biased from my C programming background and expecting the key function description information to be in the name and not the parameters ....

Reply:
How about swapping the path and size:

```
- (void)fileWasRemovedAtPath:(NSString *)path withSize:(GDTCORStorageSizeBytes)fileSize;
- (void)fileWasAddedAtPath:(NSString *)path withSize:(GDTCORStorageSizeBytes)fileSize;
```

`withSize` could also just be `size` as well.

Side note - should we be using file URLs here instead of NSStrings?

Reply:
I like Ryan's suggestion the most so far. 

As for URLs - NSString with paths are uses all across the storage now. I think it will be better to replace them all in one shot to avoid path-to-URL conversions. Another concern with URLs is that URL is a path with URL scheme, parameters, etc. so we will probably need to make more validation on it. I would defer URLs change.